src/vs/workbench/contrib/terminal/node/terminalProfiles.ts

```diff
@@ -0,0 +1,154 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as fs from 'fs';
+import * as platform from 'vs/base/common/platform';
+import { coalesce } from 'vs/base/common/arrays';
+import { normalize, basename } from 'vs/base/common/path';
+import { enumeratePowerShellInstallations } from 'vs/base/node/powershell';
+import { getWindowsBuildNumber } from 'vs/platform/terminal/node/terminalEnvironment';
+import { ITerminalProfile } from 'vs/workbench/contrib/terminal/common/terminal';
+import * as cp from 'child_process';
+
+export interface IStatProvider {
+	stat(path: string): boolean,
+	lstat(path: string): boolean
+}
+
+export function detectAvailableShells(detectWslShells?: boolean, statProvider?: IStatProvider): Promise<ITerminalProfile[]> {
+	return platform.isWindows ? detectAvailableWindowsShells(detectWslShells, statProvider) : detectAvailableUnixShells();
+}
+
+async function detectAvailableWindowsShells(detectWslShells?: boolean, statProvider?: IStatProvider): Promise<ITerminalProfile[]> {
+	// Determine the correct System32 path. We want to point to Sysnative
+	// when the 32-bit version of VS Code is running on a 64-bit machine.
+	// The reason for this is because PowerShell's important PSReadline
+	// module doesn't work if this is not the case. See #27915.
+	const is32ProcessOn64Windows = process.env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
+	const system32Path = `${process.env['windir']}\\${is32ProcessOn64Windows ? 'Sysnative' : 'System32'}`;
+
+	let useWSLexe = false;
+
+	if (getWindowsBuildNumber() >= 16299) {
+		useWSLexe = true;
+	}
+
+	let expectedLocations: { [key: string]: string[] } = {
+		'Command Prompt': [`${system32Path}\\cmd.exe`],
+		'WSL Bash': [`${system32Path}\\${useWSLexe ? 'wsl.exe' : 'bash.exe'}`],
+		'Git Bash': [
+			`${process.env['ProgramW6432']}\\Git\\bin\\bash.exe`,
+			`${process.env['ProgramW6432']}\\Git\\usr\\bin\\bash.exe`,
+			`${process.env['ProgramFiles']}\\Git\\bin\\bash.exe`,
+			`${process.env['ProgramFiles']}\\Git\\usr\\bin\\bash.exe`,
+			`${process.env['LocalAppData']}\\Programs\\Git\\bin\\bash.exe`,
+		],
+		'Cygwin': [
+			`${process.env['HOMEDRIVE']}\\cygwin64\\bin\\bash.exe`,
+			`${process.env['HOMEDRIVE']}\\cygwin\\bin\\bash.exe`
+		]
+	};
+
+	// Add all of the different kinds of PowerShells
+	for await (const pwshExe of enumeratePowerShellInstallations()) {
+		expectedLocations[pwshExe.displayName] = [pwshExe.exePath];
+	}
+	const promises: Promise<ITerminalProfile | undefined>[] = [];
+	Object.keys(expectedLocations).forEach(key => promises.push(validateShellPaths(key, expectedLocations[key], statProvider)));
+	const shells = await Promise.all(promises);
+	let output = '';
+	cp.exec('wsl.exe -l', (err, stdout) => {
```

Comment:
`output` might not get set by the time you  need it, use await here:

```ts
// TODO: Handle errors too
const distroOutput = new Promise<string>(r => cp.exec('wsl.exe -l', (err, stdout) => r(stdout)));
```

We may have utils that wrap exec for us.